File: src/main/java/com/epri/dss/meter/impl/SensorImpl.java
package com.epri.dss.meter.impl;

import com.epri.dss.common.Circuit;
import com.epri.dss.common.CktElement;
import com.epri.dss.common.impl.DSSClassDefs;
import com.epri.dss.common.impl.DSSGlobals;
import com.epri.dss.common.impl.Utilities;
import com.epri.dss.conversion.PCElement;
import com.epri.dss.delivery.PDElement;
import com.epri.dss.meter.Sensor;
import com.epri.dss.meter.SensorObj;
import com.epri.dss.parser.impl.Parser;
import com.epri.dss.shared.impl.CommandListImpl;

public class SensorImpl extends MeterClassImpl implements Sensor {

	public static SensorObj activeSensorObj;

	public SensorImpl() {
		super();

		className = "Sensor";
		classType = classType + DSSClassDefs.SENSOR_ELEMENT;

		defineProperties();

		String[] commands = new String[numProperties];
		System.arraycopy(propertyName, 0, commands, 0, numProperties);
		commandList = new CommandListImpl(commands);
		commandList.setAbbrevAllowed(true);
	}

	protected void defineProperties() {

		numProperties = NumPropsThisClass;
		countProperties();  // get inherited property count
		allocatePropertyArrays();

		// define property names
		propertyName[0] = "element";
		propertyName[1] = "terminal";
		propertyName[2] = "kvbase";
		propertyName[3] = "clear";
		propertyName[4] = "kVs";
		propertyName[5] = "currents";
		propertyName[6] = "kWs";
		propertyName[7] = "kvars";
		propertyName[8] = "conn";  // sensor connection
		propertyName[9] = "Deltadirection";  // +/- 1
		propertyName[10] = "%Error";  // %Error of sensor
		propertyName[11] = "Weight";  // for WLS calc
		propertyName[12] = "action";

		propertyHelp[0] = "Name (Full Object name) of element to which the Sensor is connected.";
		propertyHelp[1] = "Number of the terminal of the circuit element to which the Sensor is connected. "+
				"1 or 2, typically. Default is 1.";
		propertyHelp[2] = "Voltage base for the sensor, in kV. If connected to a 2- or 3-phase terminal, " + DSSGlobals.CRLF +
				"specify L-L voltage. For 1-phase devices specify L-N or actual 1-phase voltage. "+
				"Like many other DSS devices, default is 12.47kV.";
		propertyHelp[3] = "{ Yes | No }. Clear=Yes clears sensor values. Should be issued before putting in a new set of measurements.";
		propertyHelp[4] = "Array of Voltages (kV) measured by the voltage sensor. For Delta-connected " +
				"sensors, Line-Line voltages are expected. For Wye, Line-Neutral are expected.";
		propertyHelp[5] = "Array of Currents (amps) measured by the current sensor. Specify this or power quantities; not both.";
		propertyHelp[6] = "Array of Active power (kW) measurements at the sensor. Is converted into Currents along with q=[...]"+DSSGlobals.CRLF+
				"Will override any i=[...] specification.";
		propertyHelp[7] = "Array of Reactive power (kvar) measurements at the sensor. Is converted into Currents along with p=[...]";
		propertyHelp[8] = "Voltage sensor Connection: { wye | delta | LN | LL }.  Default is wye. Applies to voltage measurement only. "+DSSGlobals.CRLF+
				"Currents are always assumed to be line currents." + DSSGlobals.CRLF +
				"If wye or LN, voltage is assumed measured line-neutral; otherwise, line-line.";
		propertyHelp[9] ="{1 or -1}  Default is 1:  1-2, 2-3, 3-1.  For reverse rotation, enter -1. Any positive or negative entry will suffice.";
		propertyHelp[10] ="Assumed percent error in the measurement. Default is 1.";
		propertyHelp[11] ="Weighting factor: Default is 1.";
		propertyHelp[12] ="NOT IMPLEMENTED.Action options: "+DSSGlobals.CRLF+"SQERROR: Show square error of the present value of the monitored terminal  "+DSSGlobals.CRLF+
				"quantity vs the sensor value. Actual values - convert to per unit in calling program.  "+DSSGlobals.CRLF+
				"Value reported in result window/result variable.";

		activeProperty = Sensor.NumPropsThisClass - 1;
		super.defineProperties();  // add defs of inherited properties to bottom of list
	}

	@Override
	public int newObject(String objName) {
		DSSGlobals.activeCircuit.setActiveCktElement(new SensorObjImpl(this, objName));
		return addObjectToList(DSSGlobals.activeDSSObject);
	}

	@Override
	public int edit() {
		Parser parser = Parser.getInstance();

		// continue parsing with contents of parser
		activeSensorObj = (SensorObj) elementList.getActive();
		DSSGlobals.activeCircuit.setActiveCktElement(activeSensorObj);

		int result = 0;
		boolean doRecalcElementData = false;

		SensorObj as = activeSensorObj;

		int paramPointer = -1;
		String paramName = parser.getNextParam();
		String param = parser.makeString();
		while (param.length() > 0) {
			if (paramName.length() == 0) {
				paramPointer += 1;
			} else {
				paramPointer = commandList.getCommand(paramName);
			}

			if (paramPointer >= 0 && paramPointer < numProperties)
				as.setPropertyValue(paramPointer, param);

			switch (paramPointer) {
			case -1:
				DSSGlobals.doSimpleMsg("Unknown parameter \"" + paramName + "\" for object \"" + getName() +"."+ as.getName() + "\"", 661);
				break;
			case 0:
				as.setElementName(param.toLowerCase());
				break;
			case 1:
				as.setMeteredTerminal(parser.makeInteger());
				break;
			case 2:
				as.setKVBase(parser.makeDouble());
				break;
			case 3:
				as.setClearSpecified(Utilities.interpretYesNo(param));
				break;
			case 4:
				parser.parseAsVector(as.getNPhases(), as.getSensorVoltage());  // inits to zero
				break;
			case 5:
				parser.parseAsVector(as.getNPhases(), as.getSensorCurrent());  // inits to zero
				break;
			case 6:
				parser.parseAsVector(as.getNPhases(), as.getSensorKW());
				break;
			case 7:
				parser.parseAsVector(as.getNPhases(), as.getSensorKVAr());
				break;
			case 8:
				as.setConn(Utilities.interpretConnection(param));
				break;
			case 9:
				as.setDeltaDirection( as.limitToPlusMinusOne(parser.makeInteger()) );
				break;
			case 10:
				as.setPctError(parser.makeDouble());
				break;
			case 11:
				as.setWeight(parser.makeDouble());
				break;
			case 12:
				as.setAction(param);  // put sq error in global result
				break;
			default:
				// inherited parameters
				classEdit(activeSensorObj, paramPointer - Sensor.NumPropsThisClass);
				break;
			}

			switch (paramPointer) {
			case 0:
				doRecalcElementData = true;
				as.setMeteredElementChanged(true);
				break;
			case 1:
				doRecalcElementData = true;
				as.setMeteredElementChanged(true);
				break;
			case 2:
				doRecalcElementData = true;
				break;

			/* Do not recalc element data for setting of sensor quantities */
			case 3:
				if (as.isClearSpecified()) as.clearSensor();
				break;
			case 4:
				as.setVSpecified(true);
				break;
			case 5:
				as.setISpecified(true);
				break;
			case 6:
				as.setPSpecified(true);
				break;
			case 7:
				as.setQSpecified(true);
				break;
			case 8:
				doRecalcElementData = true;
				break;
			case 9:
				doRecalcElementData = true;
				break;
			}

			paramName = parser.getNextParam();
			param = parser.makeString();
		}

		if (doRecalcElementData)
			as.recalcElementData();

		return result;
	}

	/**
	 * Force all sensors in the circuit to reset.
	 */
	@Override
	public void resetAll() {
		for (SensorObj sensor : DSSGlobals.activeCircuit.getSensors())
			sensor.resetIt();
	}

	/**
	 * Force all sensors to take a sample.
	 */
	@Override
	public void sampleAll() {
		for (SensorObj sensor : DSSGlobals.activeCircuit.getSensors())
			sensor.takeSample();
	}

	/**
	 * Force all sensors to save their buffers to disk.
	 */
	@Override
	public void saveAll() {
//		for (SensorObj pSensor : DSSGlobals.activeCircuit.getSensors())
//			pSensor.save();
	}

	/**
	 * Set the hasSensorObj flag for all ckt element.
	 */
	public void setHasSensorFlag() {
		int i;
		SensorObj thisSensor;
		CktElement cktElem;

		Circuit ckt = DSSGlobals.activeCircuit;

		/* Initialize all to false */
		for (i = 0; i < ckt.getPDElements().size(); i++) {
			cktElem = ckt.getPDElements().get(i);
			cktElem.setHasSensorObj(false);
		}
		for (i = 0; i < ckt.getPCElements().size(); i++) {
			cktElem = ckt.getPCElements().get(i);
			cktElem.setHasSensorObj(false);
		}

		for (i = 0; i < ckt.getSensors().size(); i++) {
			thisSensor = ckt.getSensors().get(i);
			if (thisSensor.getMeteredElement() != null) {
				thisSensor.getMeteredElement().setHasSensorObj(true);
				if (thisSensor.getMeteredElement() instanceof PCElement) {
					((PCElement) thisSensor.getMeteredElement()).setSensorObj(thisSensor);
				} else {
					((PDElement) thisSensor.getMeteredElement()).setSensorObj(thisSensor);
				}
			}
		}
	}

	@Override
	protected int makeLike(String sensorName) {
		SensorObj otherSensor;
		int i, result = 0;

		/* See if we can find this sensor name in the present collection */
		otherSensor = (SensorObj) find(sensorName);
		if (otherSensor != null) {
			SensorObj as = activeSensorObj;

			as.setNPhases(otherSensor.getNPhases());
			as.setNConds(otherSensor.getNConds());  // force reallocation of terminal stuff

			as.setElementName(otherSensor.getElementName());
			as.setMeteredElement(otherSensor.getMeteredElement());  // target circuit element
			as.setMeteredTerminal(otherSensor.getMeteredTerminal());

			for (i = 0; i < as.getParentClass().getNumProperties(); i++)
				as.setPropertyValue(i, otherSensor.getPropertyValue(i));

			as.setBaseFrequency(otherSensor.getBaseFrequency());
		} else {
			DSSGlobals.doSimpleMsg("Error in Sensor makeLike: \"" + sensorName + "\" not found.", 662);
		}

		return result;
	}

	@Override
	public int init(int handle) {
		SensorObj sensor;
		int Result = 0;

		if (handle >= 0) {
			sensor = (SensorObj) elementList.get(handle);
			sensor.resetIt();
		} else {  // do them all
			for (int i = 0; i < elementList.size(); i++) {
				sensor = (SensorObj) elementList.get(i);
				sensor.resetIt();
			}
		}

		return Result;
	}

}

File: pydss/meter/SensorImpl.py
from dss.meter.impl.SensorObjImpl import (SensorObjImpl,)
from dss.common.impl.Utilities import (Utilities,)
from dss.meter.Sensor import (Sensor,)
from dss.common.impl.DSSGlobals import (DSSGlobals,)
from dss.shared.impl.CommandListImpl import (CommandListImpl,)
from dss.meter.impl.MeterClassImpl import (MeterClassImpl,)
from dss.common.impl.DSSClassDefs import (DSSClassDefs,)
from dss.parser.impl.Parser import (Parser,)


class SensorImpl(MeterClassImpl, Sensor):
    activeSensorObj = None

    def __init__(self):
        super(SensorImpl, self)()
        self.className = 'Sensor'
        self.classType = self.classType + DSSClassDefs.SENSOR_ELEMENT
        self.defineProperties()
        commands = [None] * self.numProperties
        System.arraycopy(self.propertyName, 0, commands, 0, self.numProperties)
        self.commandList = CommandListImpl(commands)
        self.commandList.setAbbrevAllowed(True)

    def defineProperties(self):
        self.numProperties = self.NumPropsThisClass
        self.countProperties()
        # get inherited property count
        self.allocatePropertyArrays()
        # define property names
        self.propertyName[0] = 'element'
        self.propertyName[1] = 'terminal'
        self.propertyName[2] = 'kvbase'
        self.propertyName[3] = 'clear'
        self.propertyName[4] = 'kVs'
        self.propertyName[5] = 'currents'
        self.propertyName[6] = 'kWs'
        self.propertyName[7] = 'kvars'
        self.propertyName[8] = 'conn'
        # sensor connection
        self.propertyName[9] = 'Deltadirection'
        # +/- 1
        self.propertyName[10] = '%Error'
        # %Error of sensor
        self.propertyName[11] = 'Weight'
        # for WLS calc
        self.propertyName[12] = 'action'
        self.propertyHelp[0] = 'Name (Full Object name) of element to which the Sensor is connected.'
        self.propertyHelp[1] = 'Number of the terminal of the circuit element to which the Sensor is connected. ' + '1 or 2, typically. Default is 1.'
        self.propertyHelp[2] = 'Voltage base for the sensor, in kV. If connected to a 2- or 3-phase terminal, ' + DSSGlobals.CRLF + 'specify L-L voltage. For 1-phase devices specify L-N or actual 1-phase voltage. ' + 'Like many other DSS devices, default is 12.47kV.'
        self.propertyHelp[3] = '{ Yes | No }. Clear=Yes clears sensor values. Should be issued before putting in a new set of measurements.'
        self.propertyHelp[4] = 'Array of Voltages (kV) measured by the voltage sensor. For Delta-connected ' + 'sensors, Line-Line voltages are expected. For Wye, Line-Neutral are expected.'
        self.propertyHelp[5] = 'Array of Currents (amps) measured by the current sensor. Specify this or power quantities; not both.'
        self.propertyHelp[6] = 'Array of Active power (kW) measurements at the sensor. Is converted into Currents along with q=[...]' + DSSGlobals.CRLF + 'Will override any i=[...] specification.'
        self.propertyHelp[7] = 'Array of Reactive power (kvar) measurements at the sensor. Is converted into Currents along with p=[...]'
        self.propertyHelp[8] = 'Voltage sensor Connection: { wye | delta | LN | LL }.  Default is wye. Applies to voltage measurement only. ' + DSSGlobals.CRLF + 'Currents are always assumed to be line currents.' + DSSGlobals.CRLF + 'If wye or LN, voltage is assumed measured line-neutral; otherwise, line-line.'
        self.propertyHelp[9] = '{1 or -1}  Default is 1:  1-2, 2-3, 3-1.  For reverse rotation, enter -1. Any positive or negative entry will suffice.'
        self.propertyHelp[10] = 'Assumed percent error in the measurement. Default is 1.'
        self.propertyHelp[11] = 'Weighting factor: Default is 1.'
        self.propertyHelp[12] = 'NOT IMPLEMENTED.Action options: ' + DSSGlobals.CRLF + 'SQERROR: Show square error of the present value of the monitored terminal  ' + DSSGlobals.CRLF + 'quantity vs the sensor value. Actual values - convert to per unit in calling program.  ' + DSSGlobals.CRLF + 'Value reported in result window/result variable.'
        self.activeProperty = Sensor.NumPropsThisClass - 1
        super(SensorImpl, self).defineProperties()
        # add defs of inherited properties to bottom of list

    def newObject(self, objName):
        DSSGlobals.activeCircuit.setActiveCktElement(SensorObjImpl(self, objName))
        return self.addObjectToList(DSSGlobals.activeDSSObject)

    def edit(self):
        parser = Parser.getInstance()
        # continue parsing with contents of parser
        self.activeSensorObj = self.elementList.getActive()
        DSSGlobals.activeCircuit.setActiveCktElement(self.activeSensorObj)
        result = 0
        doRecalcElementData = False
        as_ = self.activeSensorObj
        paramPointer = 0
        paramName = parser.getNextParam()
        param = parser.makeString()
        while len(param) > 0:
            if len(paramName) == 0:
                paramPointer += 1
            else:
                paramPointer = self.commandList.getCommand(paramName)
            if paramPointer >= 0 and paramPointer < self.numProperties:
                as_.setPropertyValue(paramPointer, param)
            _0 = paramPointer
            _1 = False
            while True:
                if _0 == -1:
                    _1 = True
                    DSSGlobals.doSimpleMsg('Unknown parameter \"' + paramName + '\" for object \"' + self.getName() + '.' + as_.getName() + '\"', 661)
                    break
                if (_1 is True) or (_0 == 0):
                    _1 = True
                    as_.setElementName(param.toLowerCase())
                    break
                if (_1 is True) or (_0 == 1):
                    _1 = True
                    as_.setMeteredTerminal(parser.makeInteger())
                    break
                if (_1 is True) or (_0 == 2):
                    _1 = True
                    as_.setKVBase(parser.makeDouble())
                    break
                if (_1 is True) or (_0 == 3):
                    _1 = True
                    as_.setClearSpecified(Utilities.interpretYesNo(param))
                    break
                if (_1 is True) or (_0 == 4):
                    _1 = True
                    parser.parseAsVector(as_.getNPhases(), as_.getSensorVoltage())
                    # inits to zero
                    break
                if (_1 is True) or (_0 == 5):
                    _1 = True
                    parser.parseAsVector(as_.getNPhases(), as_.getSensorCurrent())
                    # inits to zero
                    break
                if (_1 is True) or (_0 == 6):
                    _1 = True
                    parser.parseAsVector(as_.getNPhases(), as_.getSensorKW())
                    break
                if (_1 is True) or (_0 == 7):
                    _1 = True
                    parser.parseAsVector(as_.getNPhases(), as_.getSensorKVAr())
                    break
                if (_1 is True) or (_0 == 8):
                    _1 = True
                    as_.setConn(Utilities.interpretConnection(param))
                    break
                if (_1 is True) or (_0 == 9):
                    _1 = True
                    as_.setDeltaDirection(as_.limitToPlusMinusOne(parser.makeInteger()))
                    break
                if (_1 is True) or (_0 == 10):
                    _1 = True
                    as_.setPctError(parser.makeDouble())
                    break
                if (_1 is True) or (_0 == 11):
                    _1 = True
                    as_.setWeight(parser.makeDouble())
                    break
                if (_1 is True) or (_0 == 12):
                    _1 = True
                    as_.setAction(param)
                    # put sq error in global result
                    break
                if True:
                    _1 = True
                    self.classEdit(self.activeSensorObj, paramPointer - Sensor.NumPropsThisClass)
                    break
                break
            _2 = paramPointer
            _3 = False
            while True:
                if _2 == 0:
                    _3 = True
                    doRecalcElementData = True
                    as_.setMeteredElementChanged(True)
                    break
                if (_3 is True) or (_2 == 1):
                    _3 = True
                    doRecalcElementData = True
                    as_.setMeteredElementChanged(True)
                    break
                if (_3 is True) or (_2 == 2):
                    _3 = True
                    doRecalcElementData = True
                    break
                    # Do not recalc element data for setting of sensor quantities
                if (_3 is True) or (_2 == 3):
                    _3 = True
                    if as_.isClearSpecified():
                        as_.clearSensor()
                    break
                if (_3 is True) or (_2 == 4):
                    _3 = True
                    as_.setVSpecified(True)
                    break
                if (_3 is True) or (_2 == 5):
                    _3 = True
                    as_.setISpecified(True)
                    break
                if (_3 is True) or (_2 == 6):
                    _3 = True
                    as_.setPSpecified(True)
                    break
                if (_3 is True) or (_2 == 7):
                    _3 = True
                    as_.setQSpecified(True)
                    break
                if (_3 is True) or (_2 == 8):
                    _3 = True
                    doRecalcElementData = True
                    break
                if (_3 is True) or (_2 == 9):
                    _3 = True
                    doRecalcElementData = True
                    break
                break
            paramName = parser.getNextParam()
            param = parser.makeString()
        if doRecalcElementData:
            as_.recalcElementData()
        return result

    def resetAll(self):
        """Force all sensors in the circuit to reset."""
        for sensor in DSSGlobals.activeCircuit.getSensors():
            sensor.resetIt()

    def sampleAll(self):
        """Force all sensors to take a sample."""
        for sensor in DSSGlobals.activeCircuit.getSensors():
            sensor.takeSample()

    def saveAll(self):
        """Force all sensors to save their buffers to disk."""
        # for (SensorObj pSensor : DSSGlobals.activeCircuit.getSensors())
        # pSensor.save();
        pass

    def setHasSensorFlag(self):
        """Set the hasSensorObj flag for all ckt element."""
        ckt = DSSGlobals.activeCircuit
        # Initialize all to false
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < len(ckt.getPDElements())):
                break
            cktElem = ckt.getPDElements().get(i)
            cktElem.setHasSensorObj(False)
        _1 = True
        i = 0
        while True:
            if _1 is True:
                _1 = False
            else:
                i += 1
            if not (i < len(ckt.getPCElements())):
                break
            cktElem = ckt.getPCElements().get(i)
            cktElem.setHasSensorObj(False)
        _2 = True
        i = 0
        while True:
            if _2 is True:
                _2 = False
            else:
                i += 1
            if not (i < len(ckt.getSensors())):
                break
            thisSensor = ckt.getSensors().get(i)
            if thisSensor.getMeteredElement() is not None:
                thisSensor.getMeteredElement().setHasSensorObj(True)
                if isinstance(thisSensor.getMeteredElement(), PCElement):
                    thisSensor.getMeteredElement().setSensorObj(thisSensor)
                else:
                    thisSensor.getMeteredElement().setSensorObj(thisSensor)

    def makeLike(self, sensorName):
        result = 0
        # See if we can find this sensor name in the present collection
        otherSensor = self.find(sensorName)
        if otherSensor is not None:
            as_ = self.activeSensorObj
            as_.setNPhases(otherSensor.getNPhases())
            as_.setNConds(otherSensor.getNConds())
            # force reallocation of terminal stuff
            as_.setElementName(otherSensor.getElementName())
            as_.setMeteredElement(otherSensor.getMeteredElement())
            # target circuit element
            as_.setMeteredTerminal(otherSensor.getMeteredTerminal())
            _0 = True
            i = 0
            while True:
                if _0 is True:
                    _0 = False
                else:
                    i += 1
                if not (i < as_.getParentClass().getNumProperties()):
                    break
                as_.setPropertyValue(i, otherSensor.getPropertyValue(i))
            as_.setBaseFrequency(otherSensor.getBaseFrequency())
        else:
            DSSGlobals.doSimpleMsg('Error in Sensor makeLike: \"' + sensorName + '\" not found.', 662)
        return result

    def init(self, handle):
        Result = 0
        if handle >= 0:
            sensor = self.elementList.get(handle)
            sensor.resetIt()
        else:
            # Do 'em all
            _0 = True
            i = 0
            while True:
                if _0 is True:
                    _0 = False
                else:
                    i += 1
                if not (i < len(self.elementList)):
                    break
                sensor = self.elementList.get(i)
                sensor.resetIt()
        return Result
